core/types/block_ext.go

```diff
@@ -7,50 +7,117 @@ import (
 	"math/big"
 
 	"github.com/ava-labs/libevm/common"
+	ethtypes "github.com/ava-labs/libevm/core/types"
+	"github.com/ava-labs/libevm/rlp"
 )
 
-func (b *Block) WithExtData(version uint32, extdata *[]byte) *Block {
-	b.version = version
-	b.setExtDataHelper(extdata, false)
-	return b
+// BlockBodyExtra is a struct containing extra fields used by Avalanche
+// in the [Block] and [Body].
+type BlockBodyExtra struct {
+	Version uint32
+	ExtData *[]byte
+
+	// RLP block and body field removed from geth:
+	// - Withdrawals []*Withdrawal
 }
 
-func (b *Block) setExtDataHelper(data *[]byte, recalc bool) {
-	if data == nil {
-		b.setExtData(nil, recalc)
-		return
+// Copy deep copies the [BlockBodyExtra] `b` and returns it.
+// It is notably used in the following functions:
+// - [Block]'s Body() method
+// - [Block]'s With*() methods
+func (b *BlockBodyExtra) Copy() *BlockBodyExtra {
+	cpy := *b
+	if b.ExtData == nil {
+		data := make([]byte, 0)
+		cpy.ExtData = &data
```

Comment:
why do we copy nil as empty slice?

Reply:
Let's keep it as it is and limit possibility of problems introduced in the libevm branch.
We can eventually refactor later, but anyway changing this probably doesn't really matter i.e. performance-wise.

Reply:
We'll be re-running the full chain history and also running a canary for a while to confirm.

Reply:
Let's YOLO it then, changed to just

```go
	if b.ExtData != nil {
		data := slices.Clone(*b.ExtData)
		cpy.ExtData = &data
	}
```